# Prediction of New Onset Atrial Fibrillation Using Routinely Reported 12-Lead ECG Variables and Electronic Health Data

## Table of Contents

- Background
- Research Question
- Data Dictionary
- Exploratory Data Analysis
- References

## Background

Atrial fibrillation (AF) is the most common irregular heart rhythm and is often described as a cardiovascular epidemic of the 21st century. It affects approximately 1 in 3 people over the age of 45 *(Kornej et al., 2020; Linz et al., 2024)*. One of the most dangerous complications of AF is the formation of blood clots in the heart, which can travel to the brain and cause strokes. Individuals with AF are 4 to 5 times more likely to experience a stroke *(Kornej et al., 2020; Healey et al., 2012)*. Predicting who might develop AF is important so that doctors can initiate preventive treatments, such as blood-thinning medications, at an early stage.

Several risk scores have been developed for the prediction of AF using traditional statistical models, such as the C2HEST *(Li et al., 2019)* and CHARGE-AF *(Alonso et al., 2013)* scores, with modest performance in validation datasets (C-index 0.59-0.73). Additionally, some of these scores have been derived for restricted ethnicity groups.

Recently, machine learning (ML) algorithms have been explored for this task and have shown improved predictive performance. One study used data from patients, electronic health records (EHR), and heart MRI scans to predict AF over five years with a prediction accuracy (C-index) of 0.78 *(Dykstra et al., 2022)*. Another model called FIND-AF was developed in the UK using routine EHR data to predict new cases of AF within six months, with a high accuracy (ROC-AUC of 0.82) *(Nadarajah et al., 2023)*.

There is still a need for better AF prediction models that work well for all types of patients and clinical settings. **It is not yet clear if including ECG (heart rhythm test) data can make predictions more accurate than using only EHR data.**

## Research Question

Can a **risk prediction model** be developed using a **large repository of synthetic patient health data**, including **12-lead ECG** and **electronic health record (EHR) variables** from patients in **Southern Alberta** with suspected or known **cardiovascular disease**, to accurately predict the **future occurrence of new-onset atrial fibrillation (AF)** for individual patients?

**Study cohort**: A synthetic dataset of about 100,000 patients without a history of AF. These patients had a baseline ECG done between January 2010 and January 2023 and were followed for at least 12 months. Patients with current or past AF/flutter were excluded based on their baseline ECG and records from continuous ECG monitoring (Holter), diagnostic codes (ICD-10-CA), or procedure codes related to AF/flutter treatment. This synthetic data was created using a random sample of approximately 100,000 patients from the Cardiovascular Imaging Registry of Calgary (CIROC).

**Outcome of interest**: New-onset future AF/flutter detected by any follow-up ECG, continuous ambulatory ECG monitoring (Holter), ICD-10-CA code, or procedural code for AF/flutter intervention.

In [3]:
import pandas as pd

## Data Dictionary

In [5]:
data_dictionary_df = pd.read_csv('../data/data_dictionary.csv')
data_dictionary_df.head()

Variable name Section category Variable category  \
0                  patient_id           System          Tracking   
1  demographics_age_index_ecg     Demographics               Age   
2      demographics_birth_sex     Demographics               Sex   
3          hypertension_icd10     Cardiac Risk      Hypertension   
4           diabetes_combined     Cardiac Risk          Diabetes   

  Variable type                                         Definition cat_1  \
0     alpha_num  Randomly generated 9-digit alpha-numeric ident...   NaN   
1       numeric  Chronological age at time of referenced index ...   NaN   
2   categorical                              Sex assigned at birth  Male   
3       boolean  ICD-10 coding of hypertension in either DAD or...    No   
4       boolean  Documented presence of hyperglycaemic state in...    No   

    cat_2 cat_3 cat_4 cat_5  ... cat_12 cat_13 cat_14 cat_14.1 cat_15 cat_16  \
0     NaN   NaN   NaN   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN   
1     NaN   NaN   NaN   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN   
2  Female   NaN   NaN   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN   
3     Yes   NaN   NaN   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN   
4     Yes   NaN   NaN   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN   

   cat_17  cat_18  cat_19  cat_20  
0     NaN     NaN     NaN     NaN  
1     NaN     NaN     NaN     NaN  
2     NaN     NaN     NaN     NaN  
3     NaN     NaN     NaN     NaN  
4     NaN     NaN     NaN     NaN  

[5 rows x 26 columns]

In [6]:
data_dictionary_df['Section category'].value_counts()

Section category
Medications                    54
Laboratory                     50
Disease - Non-CV               13
Prior cardiovascular events    11
Disease - CV                   10
ECG                             6
Future outcomes                 5
Prior procedures - CV           4
Cardiac Risk                    3
Devices - CV                    3
Demographics                    2
System                          1
Name: count, dtype: int64

**System:**

- `patient_id`: Once duplicates are addressed, `patient_id` can be safely removed if it has no predictive value.

In [8]:
data_dictionary_df[data_dictionary_df['Section category'] == 'System']

Variable name Section category Variable category Variable type  \
0    patient_id           System          Tracking     alpha_num   

                                          Definition cat_1 cat_2 cat_3 cat_4  \
0  Randomly generated 9-digit alpha-numeric ident...   NaN   NaN   NaN   NaN   

  cat_5  ... cat_12 cat_13 cat_14 cat_14.1 cat_15 cat_16  cat_17  cat_18  \
0   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   

   cat_19  cat_20  
0     NaN     NaN  

[1 rows x 26 columns]

**Demographics:**

- `demographics_age_index_ecg`: Strong predictor; AF risk increases significantly with age due to cumulative cardiovascular changes.
- `demographics_birth_sex`: Captures sex-specific differences in AF risk and outcomes; men have higher incidence, women may have worse outcomes.

In [10]:
data_dictionary_df[data_dictionary_df['Section category'] == 'Demographics']

Variable name Section category Variable category  \
1  demographics_age_index_ecg     Demographics               Age   
2      demographics_birth_sex     Demographics               Sex   

  Variable type                                         Definition cat_1  \
1       numeric  Chronological age at time of referenced index ...   NaN   
2   categorical                              Sex assigned at birth  Male   

    cat_2 cat_3 cat_4 cat_5  ... cat_12 cat_13 cat_14 cat_14.1 cat_15 cat_16  \
1     NaN   NaN   NaN   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN   
2  Female   NaN   NaN   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN   

   cat_17  cat_18  cat_19  cat_20  
1     NaN     NaN     NaN     NaN  
2     NaN     NaN     NaN     NaN  

[2 rows x 26 columns]

**Cardiac Risk:**

- `hypertension_icd10`: Hypertension is a major modifiable risk factor for AF due to its role in promoting structural and electrical remodeling of the heart.
- `diabetes_combined`: Diabetes contributes to AF risk through systemic inflammation, oxidative stress, and cardiac remodeling.
- `dyslipidemia_combined`: Dyslipidemia indirectly affects AF risk via its contribution to atherosclerosis and cardiovascular disease.

In [12]:
data_dictionary_df[data_dictionary_df['Section category'] == 'Cardiac Risk']

Variable name Section category Variable category Variable type  \
3     hypertension_icd10     Cardiac Risk      Hypertension       boolean   
4      diabetes_combined     Cardiac Risk          Diabetes       boolean   
5  dyslipidemia_combined     Cardiac Risk      Dyslipidemia       boolean   

                                          Definition cat_1 cat_2 cat_3 cat_4  \
3  ICD-10 coding of hypertension in either DAD or...    No   Yes   NaN   NaN   
4  Documented presence of hyperglycaemic state in...    No   Yes   NaN   NaN   
5  Documented presence of dyslipidemia (treated o...    No   Yes   NaN   NaN   

  cat_5  ... cat_12 cat_13 cat_14 cat_14.1 cat_15 cat_16  cat_17  cat_18  \
3   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
4   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
5   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   

   cat_19  cat_20  
3     NaN     NaN  
4     NaN     NaN  
5     NaN     NaN  

[3 rows x 26 columns]

**Disease - CV:**

- `dcm_icd10`: Dilated cardiomyopathy increases AF risk due to structural heart changes and impaired ventricular function.
- `hcm_icd10`: Hypertrophic cardiomyopathy predisposes to AF through left atrial enlargement and fibrosis.
- `arvc_icd10`: Arrhythmogenic cardiomyopathy promotes AF via fibrofatty infiltration and arrhythmias in the right ventricle.
- `amyloid_cardiac_icd10`: Cardiac amyloidosis heightens AF risk due to atrial infiltration and stiffness.
- `myocarditis_icd10_prior`: Acute myocarditis can trigger AF through inflammation and myocardial scarring.
- `pericarditis_icd10_prior`: Acute pericarditis may increase AF risk via pericardial inflammation and atrial irritation.
- `bav_icd10`: Bicuspid aortic valve is linked to AF through associated valvular and hemodynamic abnormalities.
- `aortic_insufficiency_icd10`: Aortic insufficiency contributes to AF risk via volume overload and left atrial dilation.
- `aortic_dilation_icd10`: Aortic dilation/aneurysm raises AF risk through mechanical effects on adjacent cardiac structures.
- `aortic_dissection_icd10_prior`: Aortic dissection may indirectly increase AF risk via associated acute hemodynamic stress.

In [14]:
data_dictionary_df[data_dictionary_df['Section category'] == 'Disease - CV']

Variable name Section category  \
6                       dcm_icd10     Disease - CV   
7                       hcm_icd10     Disease - CV   
8                      arvc_icd10     Disease - CV   
9           amyloid_cardiac_icd10     Disease - CV   
10        myocarditis_icd10_prior     Disease - CV   
11       pericarditis_icd10_prior     Disease - CV   
12                      bav_icd10     Disease - CV   
13     aortic_insufficiency_icd10     Disease - CV   
14          aortic_dilation_icd10     Disease - CV   
15  aortic_dissection_icd10_prior     Disease - CV   

                   Variable category Variable type  \
6             Dilated cardiomyopathy       boolean   
7        Hypertrophic cardiomyopathy       boolean   
8      Arrhythmogenic cardiomyopathy       boolean   
9                    Cardiac amyloid       boolean   
10               Myocarditis - acute       boolean   
11              Pericarditis - acute       boolean   
12             Bicuspid aortic valve       boolean   
13  Aortic insufficiency (?moderate)       boolean   
14        Aortic dilation / aneurysm       boolean   
15                 Aortic dissection       boolean   

                                           Definition cat_1 cat_2 cat_3 cat_4  \
6   ICD-10 coded presence of dilated cardiomyopath...    No   Yes   NaN   NaN   
7   ICD-10 coded presence of hypertrophic cardiomy...    No   Yes   NaN   NaN   
8   ICD-10 coded presence of arrhythmogenic right ...    No   Yes   NaN   NaN   
9   ICD-10 coded presence of cardiac amyloidosis a...    No   Yes   NaN   NaN   
10  ICD-10 coded presence of acute myocarditis at ...    No   Yes   NaN   NaN   
11  ICD-10 coded presence of acute pericarditis at...    No   Yes   NaN   NaN   
12  ICD-10 coded presence of bicuspid aortic valve...    No   Yes   NaN   NaN   
13  ICD-10 coded presence of aortic insufficiency ...    No   Yes   NaN   NaN   
14  ICD-10 coded presence of aortic dilatation/ane...    No   Yes   NaN   NaN   
15  ICD-10 coded presence of aortic dissection at ...    No   Yes   NaN   NaN   

   cat_5  ... cat_12 cat_13 cat_14 cat_14.1 cat_15 cat_16  cat_17  cat_18  \
6    NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
7    NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
8    NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
9    NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
10   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
11   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
12   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
13   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
14   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
15   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   

    cat_19  cat_20  
6      NaN     NaN  
7      NaN     NaN  
8      NaN     NaN  
9      NaN     NaN  
10     NaN     NaN  
11     NaN     NaN  
12     NaN     NaN  
13     NaN     NaN  
14     NaN     NaN  
15     NaN     NaN  

[10 rows x 26 columns]

**Disease - Non-CV**:

- `amyloid_any_icd10`: Systemic amyloidosis can indirectly increase AF risk through multi-organ involvement, including cardiac effects.
- `copd_icd10`: COPD heightens AF risk by promoting hypoxia, systemic inflammation, and pulmonary hypertension.
- `hyperthyroid_icd10`: Hyperthyroidism is a direct trigger for AF due to increased metabolic rate and atrial excitability.
- `hypothyroid_icd10`: Hypothyroidism can contribute to AF risk via systemic effects like hypertension and diastolic dysfunction.
- `obstructive_sleep_apnea_icd10`: Strongly associated with AF due to intermittent hypoxia, atrial remodeling, and autonomic dysfunction.
- `pulmonary_htn_icd10`: Pulmonary hypertension increases AF risk by causing right atrial enlargement and strain.
- `rheumatoid_arthritis_icd10`: RA raises AF risk through chronic systemic inflammation and potential cardiovascular involvement.
- `sle_icd10`: SLE predisposes to AF due to systemic inflammation, autoimmune damage, and thrombotic risks.
- `ckd_icd10`: Chronic kidney disease promotes AF risk via uremic toxins, systemic inflammation, and electrolyte imbalances.
- `pad_icd10`: Peripheral arterial disease is a marker of systemic atherosclerosis, indirectly linked to AF risk.
- `sarcoid_icd10`: Sarcoidosis contributes to AF risk through granulomatous infiltration and potential cardiac involvement.
- `chronic_liver_disease_icd10`: Liver disease increases AF risk via systemic inflammation, coagulopathy, and hemodynamic changes.
- `cancer_any_icd10`: Cancer elevates AF risk through pro-thrombotic states, systemic inflammation, and potential cardiac involvement from treatments.

In [30]:
data_dictionary_df[data_dictionary_df['Section category'] == 'Disease - Non-CV']

Variable name  Section category  \
16               amyloid_any_icd10  Disease - Non-CV   
17                      copd_icd10  Disease - Non-CV   
18              hyperthyroid_icd10  Disease - Non-CV   
19               hypothyroid_icd10  Disease - Non-CV   
20  obstructive _sleep_apnea_icd10  Disease - Non-CV   
21             pulmonary_htn_icd10  Disease - Non-CV   
22      rheumatoid_arthritis_icd10  Disease - Non-CV   
23                       sle_icd10  Disease - Non-CV   
24                       ckd_icd10  Disease - Non-CV   
25                       pad_icd10  Disease - Non-CV   
26                   sarcoid_icd10  Disease - Non-CV   
27     chronic_liver_disease_icd10  Disease - Non-CV   
28                cancer_any_icd10  Disease - Non-CV   

                 Variable category Variable type  \
16                     Amyloidosis       boolean   
17                            COPD       boolean   
18                 Hyperthyroidism       boolean   
19                  Hypothyroidism       boolean   
20         Obstructive Sleep Apnea       boolean   
21          Pulmonary hypertension       boolean   
22            Rheumatoid arthritis       boolean   
23  Systemic Lupus Erthythamatosis       boolean   
24          Chronic kidney disease       boolean   
25       peripheral artery disease       boolean   
26                     Sarcoidosis       boolean   
27       Cirrhosis / liver disease       boolean   
28                          Cancer       boolean   

                                           Definition cat_1 cat_2 cat_3 cat_4  \
16  ICD-10 coded presence of amyloidosis at any ti...    No   Yes   NaN   NaN   
17  ICD-10 coded presence of chronic obstructive a...    No   Yes   NaN   NaN   
18  ICD-10 coded presence of hyperthyroidism at an...    No   Yes   NaN   NaN   
19  ICD-10 coded presence of hypothyroidism at any...    No   Yes   NaN   NaN   
20  ICD-10 coded presence of obstructive sleep apn...    No   Yes   NaN   NaN   
21  ICD-10 coded presence of pulmonary hypertensio...    No   Yes   NaN   NaN   
22  ICD-10 coded presence of rheumatoid arthritis ...    No   Yes   NaN   NaN   
23  ICD-10 coded presence of systemic lupus erythe...    No   Yes   NaN   NaN   
24  ICD-10 coded presence of chronic kidney diseas...    No   Yes   NaN   NaN   
25  ICD-10 coded presence of peripheral arterial d...    No   Yes   NaN   NaN   
26  ICD-10 coded presence of sarcoidosis at any ti...    No   Yes   NaN   NaN   
27  ICD-10 coded presence of liver cirrhosis/chron...    No   Yes   NaN   NaN   
28  ICD-10 coded presence of cancer at any time pr...    No   Yes   NaN   NaN   

   cat_5  ... cat_12 cat_13 cat_14 cat_14.1 cat_15 cat_16  cat_17  cat_18  \
16   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
17   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
18   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
19   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
20   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
21   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
22   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
23   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
24   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
25   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
26   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
27   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
28   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   

    cat_19  cat_20  
16     NaN     NaN  
17     NaN     NaN  
18     NaN     NaN  
19     NaN     NaN  
20     NaN     NaN  
21     NaN     NaN  
22     NaN     NaN  
23     NaN     NaN  
24     NaN     NaN  
25     NaN     NaN  
26     NaN     NaN  
27     NaN     NaN  
28     NaN     NaN  

**Prior cardiovascular events:**

- `event_cv_HF_admission_icd10_prior`: Heart failure hospitalization is a strong predictor of AF due to atrial remodeling and elevated atrial pressures.
- `event_cv_cad_acs_any_icd10_prior`: Acute coronary syndromes signal increased AF risk through ischemia-induced atrial dysfunction.
- `event_cv_cad_acs_acute_mi_icd10_prior`: Myocardial infarction directly elevates AF risk via cardiac ischemia and scar-related electrical remodeling.
- `event_cv_cad_acs_unstable_angina_icd10_prior`: Unstable angina increases AF risk as a precursor to ischemic damage and atrial strain.
- `event_cv_cad_acs_other_icd10_prior`: Other acute coronary syndromes contribute to AF risk through similar ischemic and systemic effects.
- `event_cv_ep_vt_any_icd10_prior`: Ventricular tachycardia reflects electrical instability and structural heart disease, which are associated with AF.
- `event_cv_ep_sca_survived_icd10_cci_prior`: Surviving cardiac arrest indicates severe cardiovascular dysfunction, increasing AF susceptibility.
- `event_cv_cns_stroke_any_icd10_prior`: Stroke, both ischemic and hemorrhagic, is linked to systemic risk factors and shared mechanisms for AF.
- `event_cv_cns_stroke_ischemic_icd10_prior`: Ischemic stroke reflects embolic and cardiovascular risk factors strongly tied to AF.
- `event_cv_cns_stroke_hemorrh_icd10_prior`: Hemorrhagic stroke can be linked to systemic risk factors that overlap with AF etiology.
- `event_cv_cns_TIA_icd10_prior`: TIA indicates transient ischemic episodes often associated with embolic risk factors and latent AF.

In [41]:
data_dictionary_df[data_dictionary_df['Section category'] == 'Prior cardiovascular events']

Variable name             Section category  \
29             event_cv_HF_admission_icd10_prior  Prior cardiovascular events   
30              event_cv_cad_acs_any_icd10_prior  Prior cardiovascular events   
31         event_cv_cad_acs_acute_mi_icd10_prior  Prior cardiovascular events   
32  event_cv_cad_acs_unstable_angina_icd10_prior  Prior cardiovascular events   
33            event_cv_cad_acs_other_icd10_prior  Prior cardiovascular events   
34                event_cv_ep_vt_any_icd10_prior  Prior cardiovascular events   
35      event_cv_ep_sca_survived_icd10_cci_prior  Prior cardiovascular events   
36           event_cv_cns_stroke_any_icd10_prior  Prior cardiovascular events   
37      event_cv_cns_stroke_ischemic_icd10_prior  Prior cardiovascular events   
38       event_cv_cns_stroke_hemorrh_icd10_prior  Prior cardiovascular events   
39                  event_cv_cns_TIA_icd10_prior  Prior cardiovascular events   

                 Variable category Variable type  \
29         Heart failure admission       boolean   
30         Acute coronary syndrome       boolean   
31         Acute coronary syndrome       boolean   
32         Acute coronary syndrome       boolean   
33         Acute coronary syndrome       boolean   
34         Ventricular tachycardia       boolean   
35  Survived sudden cardiac arrest       boolean   
36                          Stroke       boolean   
37                          Stroke       boolean   
38                          Stroke       boolean   
39                             TIA       boolean   

                                           Definition cat_1 cat_2 cat_3 cat_4  \
29  Heart failure hospitalization by ICD-10 coding...    No   Yes   NaN   NaN   
30  Any acute coronary syndrome (acute myocardial ...    No   Yes   NaN   NaN   
31  Acute myocardial infarction by ICD-10 coding p...    No   Yes   NaN   NaN   
32  Unstable angina by ICD-10 coding prior to inde...    No   Yes   NaN   NaN   
33  Other acute coronary syndrome by ICD-10 coding...    No   Yes   NaN   NaN   
34  Ventricular tachycardia by ICD-10 coding prior...    No   Yes   NaN   NaN   
35  Survived cardiac arrest by ICD-10 coding prior...    No   Yes   NaN   NaN   
36  Any acute stroke (ischemic or hemorrhagic) by ...    No   Yes   NaN   NaN   
37  Acute ischemic stroke by ICD-10 coding prior t...    No   Yes   NaN   NaN   
38  Acute hemorrhagic stroke by ICD-10 coding prio...    No   Yes   NaN   NaN   
39  Transient ischemic attack (TIA) by ICD-10 codi...    No   Yes   NaN   NaN   

   cat_5  ... cat_12 cat_13 cat_14 cat_14.1 cat_15 cat_16  cat_17  cat_18  \
29   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
30   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
31   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
32   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
33   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
34   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
35   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
36   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
37   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
38   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
39   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   

    cat_19  cat_20  
29     NaN     NaN  
30     NaN     NaN  
31     NaN     NaN  
32     NaN     NaN  
33     NaN     NaN  
34     NaN     NaN  
35     NaN     NaN  
36     NaN     NaN  
37     NaN     NaN  
38     NaN     NaN  
39     NaN     NaN  

[11 rows x 26 columns]

**Prior procedures - CV:**

- `pci_prior`: Prior PCI indicates coronary artery disease, which is a significant risk factor for AF due to ischemic and structural heart changes.
- `cabg_prior`: CABG reflects severe coronary artery disease, increasing AF risk via postoperative atrial remodeling and inflammation.
- `transplant_heart_cci_prior`: Heart transplantation indicates advanced cardiac disease, with AF risk linked to surgical and immunological factors.
- `lvad_cci_prior`: LVAD implantation reflects end-stage heart failure, strongly associated with AF due to atrial strain and hemodynamic changes.

In [44]:
data_dictionary_df[data_dictionary_df['Section category'] == 'Prior procedures - CV']

Variable name       Section category  \
40                   pci_prior  Prior procedures - CV   
41                  cabg_prior  Prior procedures - CV   
42  transplant_heart_cci_prior  Prior procedures - CV   
43              lvad_cci_prior  Prior procedures - CV   

                           Variable category Variable type  \
40  Percutaneous coronary intervention (PCI)       boolean   
41    Coronary artery bypass grafting (CABG)       boolean   
42                     Heart transplantation       boolean   
43                         LVAD implantation       boolean   

                                           Definition cat_1 cat_2 cat_3 cat_4  \
40  Prior PCI by CCI coding or from APPROACH datab...    No   Yes   NaN   NaN   
41  Prior coronary artery bypass grafting by CCI c...    No   Yes   NaN   NaN   
42  Prior cardiac transplantation by CCI coding at...    No   Yes   NaN   NaN   
43  Prior left ventricular assist device (LVAD) im...    No   Yes   NaN   NaN   

   cat_5  ... cat_12 cat_13 cat_14 cat_14.1 cat_15 cat_16  cat_17  cat_18  \
40   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
41   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
42   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
43   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   

    cat_19  cat_20  
40     NaN     NaN  
41     NaN     NaN  
42     NaN     NaN  
43     NaN     NaN  

[4 rows x 26 columns]

**Devices - CV:**

- `pacemaker_permanent_cci_prior`: Permanent pacemaker implantation reflects underlying conduction system disease, which is associated with increased AF risk.
- `crt_cci_prior`: Cardiac resynchronization therapy (CRT) is linked to advanced heart failure, a condition strongly associated with AF due to atrial remodeling and strain.
- `icd_cci_prior`: Implantable cardioverter defibrillator (ICD) implantation indicates severe arrhythmias or structural heart disease, both of which significantly increase AF risk.

In [47]:
data_dictionary_df[data_dictionary_df['Section category'] == 'Devices - CV']

Variable name Section category  \
44  pacemaker_permanent_cci_prior     Devices - CV   
45                  crt_cci_prior     Devices - CV   
46                  icd_cci_prior     Devices - CV   

                         Variable category Variable type  \
44                     Permanent pacemaker       boolean   
45                     Permanent pacemaker       boolean   
46  Implantable cardioverter defibrillator       boolean   

                                           Definition cat_1 cat_2 cat_3 cat_4  \
44  Prior permanent pacemaker implantation by CCI ...    No   Yes   NaN   NaN   
45  Prior cardiac resynchronization therapy (CRT) ...    No   Yes   NaN   NaN   
46  Prior internal cardioverter defibrillator (ICD...    No   Yes   NaN   NaN   

   cat_5  ... cat_12 cat_13 cat_14 cat_14.1 cat_15 cat_16  cat_17  cat_18  \
44   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
45   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   
46   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN     NaN   

    cat_19  cat_20  
44     NaN     NaN  
45     NaN     NaN  
46     NaN     NaN  

[3 rows x 26 columns]

**ECG:**

- `ecg_hr`: Heart rate provides insights into baseline cardiac function and autonomic tone, which are important for identifying AF risk.
- `ecg_pr`: PR interval reflects atrioventricular conduction; prolongation or shortening may indicate structural or electrical abnormalities linked to AF.
- `ecg_qrs`: QRS duration identifies ventricular conduction delays, which can signify underlying cardiac disease associated with AF.
- `ecg_qtc`: Corrected QT interval is important for identifying repolarization abnormalities that may predispose to arrhythmias, including AF.
- `ecg_rhythm`: Heart rhythm directly identifies arrhythmic events, with non-sinus rhythms being strong predictors of future AF.
- `ecg_qrs_morphology`: QRS morphology reflects conduction abnormalities like bundle branch blocks, which can indicate structural heart issues increasing AF risk.

In [50]:
data_dictionary_df[data_dictionary_df['Section category'] == 'ECG']

Variable name Section category Variable category Variable type  \
47              ecg_hr              ECG       12 lead ECG       numeric   
48              ecg_pr              ECG       12 lead ECG       numeric   
49             ecg_qrs              ECG       12 lead ECG       numeric   
50             ecg_qtc              ECG       12 lead ECG       numeric   
51          ecg_rhythm              ECG       12 lead ECG   categorical   
52  ecg_qrs_morphology              ECG       12 lead ECG   categorical   

                                       Definition  cat_1 cat_2     cat_3  \
47              Heart rate from index 12 lead ECG    NaN   NaN       NaN   
48    PR interval duration from index 12 lead ECG    NaN   NaN       NaN   
49    QRS complex duration from index 12 lead ECG    NaN   NaN       NaN   
50  Corrected QT interval  from index 12 lead ECG    NaN   NaN       NaN   
51            Heart rhythm from index 12 lead ECG  sinus  afib  aflutter   
52          QRS morphology from index 12 lead ECG   LBBB  RBBB      LAFB   

       cat_4                cat_5  ... cat_12 cat_13 cat_14 cat_14.1 cat_15  \
47       NaN                  NaN  ...    NaN    NaN    NaN      NaN    NaN   
48       NaN                  NaN  ...    NaN    NaN    NaN      NaN    NaN   
49       NaN                  NaN  ...    NaN    NaN    NaN      NaN    NaN   
50       NaN                  NaN  ...    NaN    NaN    NaN      NaN    NaN   
51  bigemini  wide complex rhythm  ...    NaN    NaN    NaN      NaN    NaN   
52      LPFB   Bifascicular block  ...    NaN    NaN    NaN      NaN    NaN   

   cat_16  cat_17  cat_18  cat_19  cat_20  
47    NaN     NaN     NaN     NaN     NaN  
48    NaN     NaN     NaN     NaN     NaN  
49    NaN     NaN     NaN     NaN     NaN  
50    NaN     NaN     NaN     NaN     NaN  
51    NaN     NaN     NaN     NaN     NaN  
52    NaN     NaN     NaN     NaN     NaN  

[6 rows x 26 columns]

**Laboratory:**

- `hgb_peri` & `hct_peri`: Hemoglobin and hematocrit levels indicate anemia, which can contribute to AF via increased cardiac workload and hypoxia.
- `rdw_peri`: Elevated RDW reflects variability in red cell size, linked to systemic inflammation and higher AF risk.
- `wbc_peri`: Elevated WBC counts indicate inflammation, a known contributor to AF pathogenesis.
- `plt_peri`: Platelet levels may impact clotting dynamics, relevant to AF-related thromboembolism.
- `inr_peri` & `ptt_peri`: Coagulation measures (INR, PTT) are important for understanding clotting status in patients at risk for AF complications.
- `esr_peri` & `crp_high_sensitive_peri`: Elevated ESR or CRP signals systemic inflammation, a key factor in AF development.
- `albumin_peri`: Low albumin is associated with poor overall health and increased AF risk.
- `alkaline_phophatase_peri`, `alanine_transaminase_peri`, `aspartate_transaminase_peri`, `bilirubin_total_peri`, `bilirubin_direct_peri`, `bilirubin_indirect_peri`: Liver function markers are relevant for systemic metabolic health, indirectly impacting AF risk.
- `urea_peri`, `creatinine_peri`, `uric_acid_peri`, `urine_alb_cr_ratio_peri`: Kidney function markers highlight metabolic stress and fluid balance issues, which influence AF.
- `sodium_peri`, `potassium_peri`, `chloride_peri`: Electrolyte imbalances directly affect cardiac electrophysiology and AF risk.
- `ck_peri`, `troponin_i_peri_closest`, `troponin_i_peri_highest`, `troponin_t_hs_peri_closest`, `troponin_t_hs_peri_highest`: Cardiac enzymes indicate myocardial stress or injury, strongly linked to AF risk.
- `NTproBNP_peri_closest`, `NTproBNP_peri_highest`: NT-proBNP levels reflect cardiac strain, a strong predictor of AF.
- `glucose_fasting_peri_closest`, `glucose_fasting_peri_highest`, `glucose_postprandial_peri_closest`, `glucose_postprandial_peri_highest`, `glucose_random_peri_closest`, `glucose_random_peri_highest`, `hga1c_peri_closest`, `hga1c_peri_highest`: Glucose and HbA1c levels identify diabetes and metabolic dysfunction, major AF risk factors.
- `tchol_peri_closest`, `tchol_peri_highest`, `ldl_peri_closest`, `ldl_peri_highest`, `hdl_peri_closest`, `hdl_peri_lowest`, `tg_peri_closest`, `tg_peri_highest`: Lipid levels signal cardiovascular risk and systemic inflammation, impacting AF development.
- `iron_peri`, `tibc_peri`, `ferritin_peri`: Iron studies reveal anemia or overload, both of which can influence AF via systemic effects.
- `tsh_peri`: Abnormal TSH levels indicate thyroid dysfunction, a well-known trigger for AF.

In [53]:
data_dictionary_df[data_dictionary_df['Section category'] == 'Laboratory']

Variable name Section category     Variable category  \
53                            hgb_peri       Laboratory                   CBC   
54                            hct_peri       Laboratory                   CBC   
55                            rdw_peri       Laboratory                   CBC   
56                            wbc_peri       Laboratory                   CBC   
57                            plt_peri       Laboratory                   CBC   
58                            inr_peri       Laboratory           Coagulation   
59                            ptt_peri       Laboratory           Coagulation   
60                            esr_peri       Laboratory  Inflammatory markers   
61             crp_high_sensitive_peri       Laboratory  Inflammatory markers   
62                        albumin_peri       Laboratory        Liver function   
63            alkaline_phophatase_peri       Laboratory        Liver function   
64           alanine_transaminase_peri       Laboratory        Liver function   
65         aspartate_transaminase_peri       Laboratory        Liver function   
66                bilirubin_total_peri       Laboratory        Liver function   
67               bilirubin_direct_peri       Laboratory        Liver function   
68             bilirubin_indirect_peri       Laboratory        Liver function   
69                           urea_peri       Laboratory       Kidney function   
70                     creatinine_peri       Laboratory       Kidney function   
71                      uric_acid_peri       Laboratory       Kidney function   
72             urine_alb_cr_ratio_peri       Laboratory       Kidney function   
73                         sodium_peri       Laboratory          Electrolytes   
74                      potassium_peri       Laboratory          Electrolytes   
75                       chloride_peri       Laboratory          Electrolytes   
76                             ck_peri       Laboratory       Cardiac enzymes   
77             troponin_i_peri_closest       Laboratory       Cardiac enzymes   
78             troponin_i_peri_highest       Laboratory       Cardiac enzymes   
79          troponin_t_hs_peri_closest       Laboratory       Cardiac enzymes   
80          troponin_t_hs_peri_highest       Laboratory       Cardiac enzymes   
81               NTproBNP_peri_closest       Laboratory              NTproBNP   
82               NTproBNP_peri_highest       Laboratory              NTproBNP   
83        glucose_fasting_peri_closest       Laboratory              Diabetes   
84        glucose_fasting_peri_highest       Laboratory              Diabetes   
85   glucose_postprandial_peri_closest       Laboratory              Diabetes   
86   glucose_postprandial_peri_highest       Laboratory              Diabetes   
87         glucose_random_peri_closest       Laboratory              Diabetes   
88         glucose_random_peri_highest       Laboratory              Diabetes   
89                  hga1c_peri_closest       Laboratory              Diabetes   
90                  hga1c_peri_highest       Laboratory              Diabetes   
91                  tchol_peri_closest       Laboratory                Lipids   
92                  tchol_peri_highest       Laboratory                Lipids   
93                    ldl_peri_closest       Laboratory                Lipids   
94                    ldl_peri_highest       Laboratory                Lipids   
95                    hdl_peri_closest       Laboratory                Lipids   
96                     hdl_peri_lowest       Laboratory                Lipids   
97                     tg_peri_closest       Laboratory                Lipids   
98                     tg_peri_highest       Laboratory                Lipids   
99                           iron_peri       Laboratory          Iron studies   
100                          tibc_peri       Laboratory          Iron studies   
101                      ferritin_peri       Laboratory 

**Medications:**

**Anti-Platelet/Anti-Coagulant Medications**

- `anti_platelet_oral_asa_peri` & `anti_platelet_oral_non_asa_any_peri`: Aspirin and non-ASA anti-platelets reduce thrombotic risk but may indicate a history of cardiovascular events, indirectly linked to AF risk.
- `anti_coagulant_oral_any_peri`: Use of anticoagulants suggests a high-risk thromboembolic profile, often coexisting with AF.

**Anti-Anginal Medications**
- `nitrates_any_peri` to `nitrates_dinitrates_peri` & `ranolazine_peri`: Use reflects ischemic heart disease, which is a major AF risk factor due to myocardial stress.

**ACE/ARB/Entresto Medications**
- `arb_peri`, `acei_peri`, `arni_entresto_peri`: These medications indicate hypertension or heart failure, both strongly associated with AF.

**Beta Blockers and Heart Failure Medications**
- `beta_blocker_any_peri`, `ivabradine_peri`: These suggest underlying cardiac arrhythmias or heart failure, which contribute to AF risk.

**Calcium Channel Blockers (CCBs)**
- `ccb_any_peri`, `ccb_dihydro_peri`, `ccb_non_dihydro_peri`: CCB use reflects hypertension or arrhythmias, both relevant to AF development.

**Diuretics**
- `diuretic_loop_peri` to `diuretic_vasopressin_antagonist_peri`: Diuretic use suggests fluid overload or heart failure, contributing to atrial stretch and AF risk.

**Anti-Arrhythmic Medications**
- `anti_arrhythmic_any_peri` to `anti_arrhythmic_disopyramide_peri`: Use reflects prior arrhythmias, directly linked to recurrent or new-onset AF.

**Digoxin and Myosin Inhibitors**
- `digoxin_peri`: Indicates rate control for arrhythmias, commonly used in AF management.
- `mavacamten_peri`: Suggests underlying hypertrophic cardiomyopathy, which increases AF risk.

**Amyloid Therapeutics**
- `amyloid_therapeutics_any_peri` to `amyloid_therapeutics_inotersen_peri`: Reflects treatment for systemic amyloidosis, a condition that can lead to atrial remodeling and AF.

**Lipid-Lowering Therapies**
- `lipid_any_peri` to `lipid_other_peri`: Lipid-lowering medications target cardiovascular disease, which indirectly increases AF risk through systemic effects.

**Glucose-Lowering Therapies**
- `glucose_lowering_any_peri` to `glucose_ohg_other_peri`: Use indicates diabetes, a significant metabolic risk factor for AF.

**Smoking Cessation Medications**
- `smoking_cessation_oral_peri`, `smoking_cessation_nicotine_replacement_peri`: Address smoking-related risks, which include cardiovascular inflammation and heightened AF risk.

In [56]:
data_dictionary_df[data_dictionary_df['Section category'] == 'Medications']

Variable name Section category  \
103                  anti_platelet_oral_asa_peri      Medications   
104          anti_platelet_oral_non_asa_any_peri      Medications   
105                 anti_coagulant_oral_any_peri      Medications   
106                            nitrates_any_peri      Medications   
107        nitrates_trinitrates_long_acting_peri      Medications   
108       nitrates_trinitrates_short_acting_peri      Medications   
109                   nitrates_mononitrates_peri      Medications   
110                     nitrates_dinitrates_peri      Medications   
111                              ranolazine_peri      Medications   
112                                     arb_peri      Medications   
113                                    acei_peri      Medications   
114                           arni_entresto_peri      Medications   
115                        beta_blocker_any_peri      Medications   
116                              ivabradine_peri      Medications   
117                                 ccb_any_peri      Medications   
118                             ccb_dihydro_peri      Medications   
119                         ccb_non_dihydro_peri      Medications   
120                           diuretic_loop_peri      Medications   
121                       diuretic_thiazide_peri      Medications   
122       diuretic_low_ceiling_non_thiazide_peri      Medications   
123                     diuretic_metolazone_peri      Medications   
124                     diuretic_indapamide_peri      Medications   
125                            diuretic_mra_peri      Medications   
126         diuretic_vasopressin_antagonist_peri      Medications   
127                     anti_arrhythmic_any_peri      Medications   
128              anti_arrhythmic_amiodarone_peri      Medications   
129            anti_arrhythmic_disopyramide_peri      Medications   
130                                 digoxin_peri      Medications   
131                              mavacamten_peri      Medications   
132                amyloid_therapeutics_any_peri      Medications   
133          amyloid_therapeutics_tafamidis_peri      Medications   
134         amyloid_therapeutics_diflunisal_peri      Medications   
135          amyloid_therapeutics_patisiran_peri      Medications   
136          amyloid_therapeutics_inotersen_peri      Medications   
137                               lipid_any_peri      Medications   
138                            lipid_statin_peri      Medications   
139                           lipid_fibrate_peri      Medications   
140                         lipid_ezetimibe_peri      Medications   
141                            lipid_PCKSK9_peri      Medications   
142                             lipid_other_peri      Medications   
143                    glucose_lowering_any_peri      Medications   
144                         glucose_insulin_peri      Medications   
145                  glucose_glp_1_agonsist_peri      Medications   
146                         glucose_ohg_any_peri      Medications   
147                   glucose_ohg_biguanide_peri      Medications   
148                   glucose_ohg_alphagluc_peri      Medications   
149                       glucose_ohg_dpp_4_peri      Medications   
150                      glucose_ohg_sglt_2_peri      Medications   
151                glucose_ohg_thiazolidine_peri      Medications   
152                 glucose_ohg_repaglinide_peri      Medications   
153                glucose_ohg_sulfonylurea_peri      Medications   
154                       glucose_ohg_other_peri      Medications   
155                  smoking_cessation_oral_peri      Medications   
156  smoking_cessation_nicotine_replacement_peri      Medications   

                          Variable category Variable type  \
103                                     ASA       boolean   
104                 Anti-platelet (non-ASA)       boolean   
105                     Oral anti-coagulant       boolea

**Future outcomes:**

- `event_all_death`: Indicates overall patient mortality, a critical outcome to assess the broader implications of AF and related interventions.
- `time_to_all_death`: Provides a temporal context for survival analysis, helping assess the timing of mortality relative to AF onset or other predictors.
- `event_cv_ep_afib_aflutter_new_icd10_post`: Primary outcome variable; directly identifies the occurrence of new-onset AF/flutter, essential for model predictions.
- `time_to_event_cv_ep_afib_aflutter_new_icd10_post`: Adds granularity by measuring the time until AF/flutter onset, useful for survival analysis and model calibration.
- `time_last_seen`: Tracks patient follow-up duration, ensuring accurate interpretation of non-events and censoring in time-to-event analyses.

In [62]:
data_dictionary_df[data_dictionary_df['Section category'] == 'Future outcomes']

Variable name Section category  \
157                                   event_all_death  Future outcomes   
158                                 time_to_all_death  Future outcomes   
159          event_cv_ep_afib_aflutter_new_icd10_post  Future outcomes   
160  time_to_event_cv_ep_afib_aflutter_new_icd10_post  Future outcomes   
161                                    time_last_seen  Future outcomes   

                  Variable category Variable type  \
157                 All cause death       boolean   
158         Time to all cause death       numeric   
159          New onset afib/flutter       boolean   
160  Time to new-onset afib/flutter       numeric   
161                  Time last seen       numeric   

                                            Definition cat_1 cat_2 cat_3  \
157  All cause death from Vital statistics Alberta ...    No   Yes   NaN   
158  Time from index 12 lead ECG to all cause death...   NaN   NaN   NaN   
159  New-onset atrial fibrillation/flutter by any o...    No   Yes   NaN   
160  Time from index 12 lead ECG to new onset atria...   NaN   NaN   NaN   
161  Time from index 12 lead ECG to when the patien...   NaN   NaN   NaN   

    cat_4 cat_5  ... cat_12 cat_13 cat_14 cat_14.1 cat_15 cat_16  cat_17  \
157   NaN   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN   
158   NaN   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN   
159   NaN   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN   
160   NaN   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN   
161   NaN   NaN  ...    NaN    NaN    NaN      NaN    NaN    NaN     NaN   

     cat_18  cat_19  cat_20  
157     NaN     NaN     NaN  
158     NaN     NaN     NaN  
159     NaN     NaN     NaN  
160     NaN     NaN     NaN  
161     NaN     NaN     NaN  

[5 rows x 26 columns]

## Exploratory Data Analysis

## References

Alonso, A., et al. (2013). Simple risk model predicts incidence of atrial fibrillation in a racially and geographically diverse population: The CHARGE‐AF consortium. *Journal of the American Heart Association, 2*(1).

Dykstra, S., et al. (2022). Machine learning prediction of atrial fibrillation in cardiovascular patients using cardiac magnetic resonance and electronic health information. *Frontiers in Cardiovascular Medicine, 9*.

Healey, J. S., et al. (2012). Subclinical atrial fibrillation and the risk of stroke. *New England Journal of Medicine, 366*(2), 120–129.

Kornej, J., Börschel, C. S., Benjamin, E. J., & Schnabel, R. B. (2020). Epidemiology of atrial fibrillation in the 21st century. *Circulation Research, 127*(1), 4–20.

Li, Y.-G., et al. (2019). A simple clinical risk score (C2HEST) for predicting incident atrial fibrillation in Asian subjects: Derivation in 471,446 Chinese subjects, with internal validation and external application in 451,199 Korean subjects. *Chest, 155*(3), 510–518.

Linz, D., et al. (2024). Atrial fibrillation: Epidemiology, screening and digital health. *The Lancet Regional Health – Europe, 37*, Article 100786.

Nadarajah, R., et al. (2023). Prediction of short-term atrial fibrillation risk using primary care electronic health records. *Heart, 109*(12), 1072–1079.